Add color modes to theming API (#1112)
primer/components: docs/content/theming.md

### System props and the `sx` prop

Some [system props](/system-props) and [`sx` prop](/overriding-styles) keys map to specific theme keys, allowing you to reference theme values in the corresponding object. For example, the `bg` prop maps to the `colors` theme key which means you can use the `bg` prop to reference values in the `colors` object:

```jsx
const theme = {
  colors: {
    bg: {
      primary: '#fff'
    }
  }
}

function App() {
  return (
    <ThemeProvider theme={theme}>
      <Box bg="bg.primary"></Box>
      <Box sx={{bg: 'bg.primary'}}></Box>
    </ThemeProvider>
  )
}
```

See the [Styled System Reference Table](https://styled-system.com/table) for a complete list of mappings.

### themeGet

```js
import {themeGet} from '@primer/components'
import styled from 'styled-components'

const Example = styled.div`
  background-color: ${themeGet('colors.bg.primary')};
`
```

### useTheme

```js
import {useTheme} from '@primer/components'

function Example() {
  const theme = useTheme()
  // theme.colors.bg.primary
}

> ```

Reviewer: I'm assuming there is an advantage of using this over `themeGet` otherwise we probably wouldn't have two ways of doing the same thing, can you call that out here?

## Color schemes

You can define multiple color schemes for your application using the `colors.schemes` and `shadows.schemes` objects:

```js
const theme = {
  colors: {
    text: '#000'
    schemes: {
      dark: {
        text: '#fff'
      },
      dark_dimmed: {
        text: '#aaa'
      }
    }
  },
  shadows: {
    small: '0 1px 0 rgba(0, 0, 0, 0.5)'
    schemes: {
      dark: {
        small: '0 1px 0 rgba(255, 255, 255, 0.5)'
      },
      dark_dimmed: {
        small: '0 1px 0 rgba(255, 255, 255, 0.25)'
      }
    }
  }
}
```

> The color scheme defined at the root of the `colors` and `shadows` objects will be accessible as `light`. The above example has three color schemes: `light`, `dark`, and `dark_dimmed`.

Reviewer: ```suggestion
The color scheme defined at the root of the `colors` and `shadows` objects can be accessed as `light`. The above example has three color schemes: `light`, `dark`, and `dark_dimmed`.
```

nit: small copy change, at first I thought you were talking about a11y

### Setting the initial color scheme

By default, Primer React uses the `light` color scheme. Use the `initialColorScheme` prop to override the default:

```jsx
import {ThemeProvider} from '@primer/components'

function App() {
  return (
    <ThemeProvider initialColorScheme="dark">
      <div>...</div>
    </ThemeProvider>
  )
}
```

> To automatically switch between color schemes according to your operating system setting, pass `initialColorScheme` an object that defines the color scheme for light and dark mode:

Reviewer: This is intentionally not the default behavior correct? Should we call that out anywhere?

---
title: Theming
---

Theming in Primer React is made possible by a theme object that allows you to define your application's colors, spacing, fonts, and more.

The structure of the [default theme object](/theme-reference) adheres to the [System UI Theme Specification](https://system-ui.com/theme) and contains the following keys:

- colors
- space
- fonts
- fontSizes
- fontWeights
- lineHeights
- borderWidths
- radii
- shadows
- breakpoints
- sizes

> 

Author: Where should we put definitions for `theme`, `color mode`, and `color scheme`?

    </ThemeProvider>
  )
}
```

Some components may break if your custom theme does not include all the same keys as the [default theme](/theme-reference). For that reason, we recommend extending the default theme using [deepmerge](https://www.npmjs.com/package/deepmerge).

## Referencing theme values

You can reference theme values in your application using [system props](/system-props), the [`sx` prop](/overriding-styles), the `themeGet` function, or the `useTheme` hook.

### System props and the `sx` prop

Some [system props](/system-props) and [`sx` prop](/overriding-styles) keys map to specific theme keys, allowing you to reference theme values in the corresponding object. For example, the `bg` prop maps to the `colors` theme key which means you can use the `bg` prop to reference values in the `colors` object:

```jsx
const theme = {
  colors: {
    bg: {
      primary: '#fff'
    }
  }
}

function App() {
  return (
    <ThemeProvider theme={theme}>
      <Box bg="bg.primary"></Box>
      <Box sx={{bg: 'bg.primary'}}></Box>
    </ThemeProvider>
  )
}
```

See the [Styled System Reference Table](https://styled-system.com/table) for a complete list of mappings.

### themeGet

> 

Author: Note to self: explain themeGet and useTheme

---
title: Theming
---

Theming in Primer React is made possible by a theme object that defines your application's colors, spacing, fonts, and more.

## ThemeProvider

To give components access to the theme object, you must add `ThemeProvider` to the root of your application:

```jsx
import {ThemeProvider} from '@primer/components'

function App() {
  return (
    <ThemeProvider>
      <div>...</div>
    </ThemeProvider>
  )
}
```

`ThemeProvider` comes with a [default theme object](/theme-reference) that includes colors, spacing, fonts, etc. for building applications at GitHub.

## Customizing the theme

> If you'd like to customize the [default theme](/theme-reference), you can pass your custom theme to `ThemeProvider` using the `theme` prop:

Author: ```suggestion
To customize the [default theme](/theme-reference), you can pass your custom theme to `ThemeProvider` using the `theme` prop:
```

    }
  }
}

function App() {
  return (
    <ThemeProvider theme={theme}>
      <Box bg="bg.primary"></Box>
      <Box sx={{bg: 'bg.primary'}}></Box>
    </ThemeProvider>
  )
}
```

See the [Styled System Reference Table](https://styled-system.com/table) for a complete list of mappings.

### themeGet

The `themeGet` function is a convienient way to reference theme values in styled-components template literals:

```js
import {themeGet} from '@primer/components'
import styled from 'styled-components'

const Example = styled.div`
  background-color: ${themeGet('colors.bg.primary')};
`
```

### useTheme

You can use the `useTheme` hook to reference theme values from inside any function component nested under the `ThemeProvider`:

```js
import {ThemeProvider, useTheme} from '@primer/components'

function Example() {
  const {theme} = useTheme()
  // theme.colors.bg.primary
}

function App() {
  return (
    <ThemeProvider>
      <Example />
    </ThemeProvider>
  )
}
```

<Note variant="warning">

Only use `useTheme` to reference theme values in places where it's not possible to use system props, the `sx` prop, or `themeGet`.

</Note>

## Color modes and color schemes

> The terms "color mode" and "color scheme" are often used interchangeably. However, in Primer React, they are two seperate (but related) concepts.

Reviewer 2: Typo:

```suggestion
The terms "color mode" and "color scheme" are often used interchangeably. However, in Primer React, they are two separate (but related) concepts.
```

The "color mode" of an application can be either `day`, `night`, or `auto` (i.e. synced with the operating system).

A "color scheme", on the other hand, is a collection of colors that can be associated with a color mode. The [default theme](/theme-reference) includes three color schemes: `light`, `dark`, and `dark_dimmed`. By default, the `light` scheme is displayed when the application is in `day` mode and the `dark` scheme is displayed in `night` mode.

### Setting the color mode

By default, Primer React is in `day` mode. To change the color mode, use the `colorMode` prop on `ThemeProvider` or the `setColorMode` function from the `useTheme` hook:

#### `colorMode` prop

```jsx
import {ThemeProvider} from '@primer/components'

function App() {
  return (
    // colorMode can be "day" (default), "night", or "auto"
    <ThemeProvider colorMode="auto">
      <div>...</div>
    </ThemeProvider>
  )
}
```

#### `setColorMode` function

```jsx
import {useTheme} from '@primer/components'

function Example() {
  const {setColorMode} = useTheme()
  return <button onClick={() => setColorMode('auto')}>Activate auto mode</button>
}
```

### Setting color schemes

> To change the color schemes associated with `day` and `night` mode, use the `dayScheme` and `nightScheme` props on `ThemeProvider` or the `setDayScheme` and `setNightScheme` functions from the `useTheme` hook:

Author: ```suggestion
To choose which color schemes will be displayed in `day` and `night` mode, use the `dayScheme` and `nightScheme` props on `ThemeProvider` or the `setDayScheme` and `setNightScheme` functions from the `useTheme` hook:
```
